[c10d] Polish NCCL PG monitor thread log message

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115888

We turned on monitor thread by default in https://github.com/pytorch/pytorch/pull/112518, and we want the error message that is displayed when the monitor kills the process to be more informative.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @wz337 @tianyu-l @wconstab @yf225

fduwjj committed:
[c10d] Polish monitor thread log message

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1213,7 +1213,10 @@ void ProcessGroupNCCL::heartbeatMonitor() {
       "be due to CUDA or NCCL calls being unexpectedly blocking, ",
       "especially when your program enters a deadlock state in watchdog "
-      "or destructors. If you see this error, please file a bug to PyTorch.");
+      "or destructors. If you don't believe this is the base, you can turn"
```
Suggest a complete rewrite (even the above part)

"ProcessGroupNCCL's watchdog was stuck for {TIMEOUT} seconds without making progress monitoring enqueued collectives.  This usually indicates a nccl/cuda API hang blocking the progress of the watchdog, and could be triggered by another thread holding the GIL inside a cuda api, or another bug.  If you suspect the watchdog is not actually stuck and a longer timeout would help, you can either increase `TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC` or disable the heartbeat monitor altogether (`TORCH_NCCL_ENABLE_MONITORING=0`).  If an increased timeout helps your job make progress, please file a bug to PyTorch."

fduwjj replied:
Sure... Let me do that..

fduwjj committed:
Update on "[c10d] Polish NCCL PG monitor thread log message"


We turned on monitor thread by default in https://github.com/pytorch/pytorch/pull/112518, and we want the error message that is displayed when the monitor kills the process to be more informative.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1210,10 +1210,16 @@ void ProcessGroupNCCL::heartbeatMonitor() {
+      "If you suspect the watchdog is not actually stuck and a longer timeout would help, ",
+      "you can either increase the timeout(TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC) to a larger value "
+      "or disable the heartbeat monitor(TORCH_NCCL_ENABLE_MONITORING=0)"
+      "If neither does not help, please file a bug to PyTorch.");
```
> If neither does not help, please file a bug

grammar isn't right (double negatie) - "if neither helps" is clearer 

but also i think the meaning is backwards- if increasing the timeout _DOES_ help, file a bug (bc our timeout value is probably wrong).  But if increasing timeout does _NOT_ help, watchdog is fine.

fduwjj replied:
Opps, it should be "If neither helps".

If increasing TIMEOUT helps, they can but not need to file an issue because maybe for that case, the timeout is too short? And if it does not help, then it's a hang case, they need to report hang anyway?

fduwjj replied:
OK, I updated this part to make it more clearer.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1210,10 +1210,16 @@ void ProcessGroupNCCL::heartbeatMonitor() {
+      "seconds without making progress in monitoring enqueued collectives. ",
+      "This typically indicates a nccl/cuda API hang blocking the watchdog, ",
+      "and could be triggered by another thread holding the GIL inside a ",
+      "cuda api, or other deadlock-prone behaviors.",
```
nit: capatalize CUDA and NCCL i guess

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1210,10 +1210,16 @@ void ProcessGroupNCCL::heartbeatMonitor() {
+      "cuda api, or other deadlock-prone behaviors.",
+      "If you suspect the watchdog is not actually stuck and a longer timeout would help, ",
+      "you can either increase the timeout(TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC) to a larger value "
+      "or disable the heartbeat monitor(TORCH_NCCL_ENABLE_MONITORING=0)"
```
missing a space before (

in 2 places (after timeout, then again after monitor)

fduwjj replied:
Actually I thought about it, I thought no space is actually preferred.. No strong opinion here.

fduwjj committed:
Update on "[c10d] Polish NCCL PG monitor thread log message"


We turned on monitor thread by default in https://github.com/pytorch/pytorch/pull/112518, and we want the error message that is displayed when the monitor kills the process to be more informative.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1210,10 +1210,17 @@ void ProcessGroupNCCL::heartbeatMonitor() {
+      "you can either increase the timeout (TORCH_NCCL_HEARTBEAT_TIMEOUT_SEC) to a larger value "
+      "or disable the heartbeat monitor (TORCH_NCCL_ENABLE_MONITORING=0)."
+      "If either of aforementioned helps, feel free to file an issue to PyTorch about the short timeout "
+      "or false positive abort; otherwise, please report a watchdog hang bug to PyTorch.");
```
> otherwise, please report a watchdog hang bug to PyTorch.

i think this part needs work, the rest looks great to me!

a watchdog hang is (usually?) not a pytorch bug.  So while it does make sense to contact us for help debugging, its not likely our bug.

"otherwise, please attempt to debug the hang by following X, Y, Z"

I think we should have a wiki for how to debug and triage such hangs.  And we can link to the wiki from this message.

fduwjj replied:
Yeah, that makes sense. But we need to help debug hang bug anyway that's why I put it in the log. (Let me add a TODO here. Maybe it's a good time for us to work on BE, documentation work now) We should include everything, from desync report, flight recorder, this monitor thread, etc.

wconstab reviewed (approved):
stamping i guess, might as well land it first, then write the wiki, and then replace the "please report a watchdog hang bug" part with "please follow the wiki to debug the hang" when we can add the wiki url.

fduwjj committed:
Update on "[c10d] Polish NCCL PG monitor thread log message"


We turned on monitor thread by default in https://github.com/pytorch/pytorch/pull/112518, and we want the error message that is displayed when the monitor kills the process to be more informative.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

wconstab commented:
ill fix the lint and merge @fduwjj

wconstab committed:
Update on "[c10d] Polish NCCL PG monitor thread log message"


We turned on monitor thread by default in https://github.com/pytorch/pytorch/pull/112518, and we want the error message that is displayed when the monitor kills the process to be more informative.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge -i
